pkg/util/volid.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2019 Ceph-CSI authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"strings"
+)
+
+/*
+VolumeIdentifier contains the elements that form a volume ID to be returned by CSI, and contains
+enough information to decompose and extract required cluster and pool information to locate the
+volume that relates to the volume ID.
+
+The volume identifier is composed as elaborated in the comment against ComposeVolID and thus,
+DecomposeVolID is the inverse of the same function.
+
+The VolumeIdentifier structure carries the following fields,
+- PoolID: 32 bit integer of the pool that the volume belongs to, where the ID comes from Ceph pool
+  identifier for the corresponding pool name.
+- EncodingVersion: Carries the version number of the encoding scheme used to encode the volume ID,
+  and is preserved for any future proofing w.r.t changes in the encoding scheme, and to retain
+  ability to parse backward compatible encodings.
+- ClusterID: Is a unique ID per cluster that the CSI instance is serving and is restricted to
+  lengths that can be accommodated in the encoding scheme.
+- ImageName: Is the on-disk image name of the CSI volume that corresponds to this volume ID.
```

Comment:
Nit: ... and snapshot ID? Perhaps ```VolumeIdentifier``` itself is a misnomer since it can represent snapshots? And since it only represents RBD-backed volumes and snapshots, should it be moved to the ```rbd``` package?

Reply:
Yes, started as a representation for volumes and the name got left behind. Will address naming and related comments in an update to this PR.

I would still like to retain this outside of rbd (i.e in util), as I foresee we would be reusing this in the CephFS plugin code as well, to generate and return Vol|SnapID for volumes from cephfs. objections to retaining these here?

Reply:
Is there a plan for statelessly encoding CephFS volumes? Does it fit here w/o a hammer? 

@batrick

Reply:
The term "volume" now is being used to refer to CephFS file systems. It is sufficient to just use the file system cluster id (fscid) as an identifier. Subvolumes are directories within a volume. Subvolumes is usually what end-users will use. Does that answer your question?

We also have the "ceph volume" API. I'm wondering what if any of this should be moved there. I regretfully haven't had much time to work on cleaning it up but one goal was to unify the concept of volumes for CephFS/RBD. This would be useful for simpler administration, presentation, usage monitoring, etc.

Reply:
@batrick So will each CephFS file system map to a single CSI StorageClass -- and then PVCs against that storage class would allocate sub-volumes (directories) within the parent FS (volume)? If that's the case, the ```ImageName``` would represent that sub-directory and could also just be programatically named using a UUID.

Reply:
I'm probably late to the party with such a question, but what about pre-provisioned volumes? The user/operator would need a separate tool to generate the ID in order to mount an image/volume (that has not been provisioned by this driver), right?

Reply:
>@batrick So will each CephFS file system map to a single CSI StorageClass -- and then PVCs against that storage class would allocate sub-volumes (directories) within the parent FS (volume)?

I believe so, yes.

> If that's the case, the ImageName would represent that sub-directory and could also just be programatically named using a UUID.

How would you map the UUID to a volume/subvolume?

Reply:
@gman0 AFAIK -- it wouldn't have been supported before. You would have needed to create a dummy PV, delete it behind the scenes, and rename your pre-provisioned image to the CSI image name.

Reply:
@batrick Good question (probably better to discuss in #224). If a Ceph FS ```volume``` maps to a k8s StorageClass, a higher-level tool (e.g. Rook) would be responsible for deploying those I'd suspect. As for how to map a k8s PersistentVolume to a CephFS ```sub-volume```, I guess it depends on what CSI Volume ID would need to encode a link back to the storage class name (to lookup the CephFS ```volume```) and the unique sub-volume name.

Reply:
This is more a detailing and questions around my (mis)understanding around CephFS and PVCs:

The current options in the [StorageClass](https://github.com/ceph/ceph-csi/blob/csi-v1.0/examples/cephfs/storageclass.yaml) for CephFS has the poolname, monitors, secrets (ID/keys).

A PVC, or a volume from the container orchestrator perspective, is a [sub-directory](https://github.com/ceph/ceph-csi/blob/209774160752ac5665051b2599e9a411496cc388/pkg/cephfs/volume.go#L53) under the (default/singleton?) CephFS instance from CephFS perspective. IOW, we do not have an option to specify different cephfs file system names (like so in the mount `-o mds_namespace=myfs`). Is this not needed? @gman0 any reason to have left this out?

The subdirectory is provided an extended attribute named `ceph.dir.layout.pool`, which is set to the pool name recieved from the StorageClass. The pool name passed in is not reused again, hence need not be encoded in the VolumeID that the plugin returns.

If we start supporting multiple cephfs instances, then we would need to specify the same in the StorageClass and also encode the same in the VolumeID returned. Is this the "file system cluster id (fscid)" being referred to? @dillaman and @batrick is this correct? From where can this fscid be fetched/read to use and what are its size/format specifications?

Reply:
Thalai, if I may...

> 
> A PVC, or a volume from the container orchestrator perspective, is a [sub-directory](https://github.com/ceph/ceph-csi/blob/209774160752ac5665051b2599e9a411496cc388/pkg/cephfs/volume.go#L53) under the (default/singleton?) CephFS instance from CephFS perspective. IOW, we do not have an option to specify different cephfs file system names (like so in the mount `-o mds_namespace=myfs`). Is this not needed? @gman0 any reason to have left this out?
> 

We'll want to add a Storageclass option to specify the filesystem within which the sub-directories/PVs will be created. Currently, sub-directories will be created only within the default filesystem (I think it's filesystem with fscid=1).


 
> The subdirectory is provided an extended attribute named `ceph.dir.layout.pool`, which is set to the pool name recieved from the StorageClass. The pool name passed in is not reused again, hence need not be encoded in the VolumeID that the plugin returns.
> 

FYI, `ceph.dir.layout.pool` xattr is used to specify the pool within which the data of the files belonging to directory will be stored. http://docs.ceph.com/docs/master/cephfs/file-layouts/#file-layouts 


> If we start supporting multiple cephfs instances, then we would need to specify the same in the StorageClass and also encode the same in the VolumeID returned. Is this the "file system cluster id (fscid)" being referred to?

Yes.

> @dillaman and @batrick is this correct? From where can this fscid be fetched/read to use and what are its size/format specifications?

You can fetch it using `ceph fs get <filesystem_name> -f json` (the value of "id" key is the fscid). 
See here,
https://github.com/rook/rook/blob/master/pkg/daemon/ceph/client/filesystem.go#L48

I think it'll be better to specify filesystem's name rather than it's ID in the StorageClass definition.

Do we want to allow the StorageClasses to create filesystems? Or do we want to create them using rook filesystem CRDs before hand
https://github.com/rook/rook/blob/master/Documentation/ceph-filesystem-crd.md ?

Reply:
> > From where can this fscid be fetched/read to use and what are its size/format specifications?
> 
> You can fetch it using `ceph fs get <filesystem_name> -f json` (the value of "id" key is the fscid).
> See here,
> https://github.com/rook/rook/blob/master/pkg/daemon/ceph/client/filesystem.go#L48

Thanks!

> 
> I think it'll be better to specify filesystem's name rather than it's ID in the StorageClass definition.

In the StorageClass it would be the name (akin to the pool name for RBD backed StorageClasses). In the returned VolID encoded we would use the fscid, so that it is agnostic to renames. (like, we want to store the pool ID in the VolumeID rather than the name, for similar rename constraints). Does this sound fine?

> 
> Do we want to allow the StorageClasses to create filesystems? Or do we expect them to be created before hand, e.g., using rook filesystem CRDs
> https://github.com/rook/rook/blob/master/Documentation/ceph-filesystem-crd.md ?

StorageClass defined filesystem names will not be created by the CSI plugin, it (CSI) expects them to be preexisting. It will only create a sub-directory within the given filesystem for corresponding volume requests. So, it is expected to be created before hand.

Reply:
> In the StorageClass it would be the name (akin to the pool name for RBD backed StorageClasses). In the returned VolID encoded we would use the fscid, so that it is agnostic to renames. (like, we want to store the pool ID in the VolumeID rather than the name, for similar rename constraints). Does this sound fine?

Yes.